tortoisebeacon/votescalc.go

```diff
@@ -0,0 +1,287 @@
+
+	tb.Log.With().Info("Calculated own first round votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesCount", fmt.Sprint(votesCount)),
+		log.String("ownFirstRoundVotes", fmt.Sprint(ownFirstRoundVotes)))
+
+	tb.calcVotesCount(epoch, round, votesCount)
+
+	tb.Log.With().Info("Calculated votes count",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesCount", fmt.Sprint(votesCount)))
+
+	ownCurrentRoundVotes := tb.calcOwnCurrentRoundVotes(epoch, round, votesCount)
+
+	tb.Log.With().Info("Calculated votes for one round",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("for", fmt.Sprint(ownCurrentRoundVotes.VotesFor)),
+		log.String("against", fmt.Sprint(ownCurrentRoundVotes.VotesAgainst)))
+
+	votesFor, votesAgainst := ownCurrentRoundVotes.Diff(ownFirstRoundVotes)
+
+	tb.Log.With().Info("Calculated votes diff for one round",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("for", fmt.Sprint(votesFor)),
+		log.String("against", fmt.Sprint(votesAgainst)))
+
+	return votesFor, votesAgainst
+}
+
+func (tb *TortoiseBeacon) countFirstRoundVotes(epoch types.EpochID) votesCountMap {
+	firstRoundInThisEpoch := epochRoundPair{
+		EpochID: epoch,
+		Round:   1,
+	}
+
+	tb.votesCache[firstRoundInThisEpoch] = make(map[p2pcrypto.PublicKey]votesSetPair)
+
+	firstRoundVotes := tb.incomingVotes[firstRoundInThisEpoch]
+	// TODO(nkryuchkov): as pointer is shared, ensure that maps are not changed
+	tb.votesCache[firstRoundInThisEpoch] = firstRoundVotes
+
+	votesCount := make(map[types.Hash32]int)
+
+	for pk, votesList := range firstRoundVotes {
+		firstRoundVotesFor := make(hashSet)
+		firstRoundVotesAgainst := make(hashSet)
+
+		for vote := range votesList.VotesFor {
+			votesCount[vote] += tb.voteWeight(pk)
+			firstRoundVotesFor[vote] = struct{}{}
+		}
+
+		for vote := range votesList.VotesAgainst {
+			votesCount[vote] -= tb.voteWeight(pk)
+			firstRoundVotesAgainst[vote] = struct{}{}
+		}
+
+		// copy to cache
+		tb.votesCache[firstRoundInThisEpoch][pk] = votesSetPair{
+			VotesFor:     firstRoundVotesFor,
+			VotesAgainst: firstRoundVotesAgainst,
+		}
+	}
+
+	tb.votesCountCache[firstRoundInThisEpoch] = make(map[types.Hash32]int)
+	for k, v := range tb.votesCache[firstRoundInThisEpoch] {
+		tb.votesCache[firstRoundInThisEpoch][k] = v
+	}
+
+	return votesCount
+}
+
+func (tb *TortoiseBeacon) calcOwnFirstRoundVotes(epoch types.EpochID, votesCount votesCountMap) votesSetPair {
+	ownFirstRoundsVotes := votesSetPair{
+		VotesFor:     make(hashSet),
+		VotesAgainst: make(hashSet),
+	}
+
+	for vote, count := range votesCount {
+		switch {
+		case count >= tb.threshold():
+			ownFirstRoundsVotes.VotesFor[vote] = struct{}{}
+		case count <= -tb.threshold():
+			ownFirstRoundsVotes.VotesAgainst[vote] = struct{}{}
+		case tb.weakCoin.Get(epoch, 1):
```

Comment:
isn't there another case which you vote for messages which has been received in time?